### PR TITLE
[Merged by Bors] - feat: The Erdős–Ko–Rado theorem

### PR DESCRIPTION
**Pull request by YaelDillies**
Co-authored-by: Bhavik Mehta <bhavik.mehta8@gmail.com>

---

- [x] depends on: #15000

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean [hunk 1]
```diff
@@ -290,5 +292,111 @@ theorem iterated_kk (h₁ : (𝒜 : Set (Finset (Fin n))).Sized r) (h₂ : 𝒞.
...
+  · rw [card_sdiff BsubC, hcard, hB, Nat.sub_sub_self hir]
+  · rwa [hB, card_attachFin, card_range, ← Nat.add_sub_assoc hir, Nat.add_sub_cancel_left]
+
 end KK
+
+/-- The **Erdős–Ko–Rado theorem**.
+
+The maximum size of an intersecting family in `α` where all sets have size `r` is bounded by
+`(card α - 1).choose (r - 1)`. This bound is sharp. -/
+theorem EKR {𝒜 : Finset (Finset (Fin n))} {r : ℕ} (h𝒜 : (𝒜 : Set (Finset (Fin n))).Intersecting)
```
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/KruskalKatona.lean:352]: We shouldn't continue with this name.

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean [hunk 2]
```diff
@@ -290,5 +292,112 @@ theorem iterated_kk (h₁ : (𝒜 : Set (Finset (Fin n))).Sized r) (h₂ : 𝒞.
   · refine ih h₁.shadow (kruskal_katona h₁ h₂ h₃) ?_
     convert h₃.shadow
 
+/-- A special case of the **Kruskal-Katona theorem** which is sometimes easier to work with.
+
+If `|𝒜| ≥ k choose r`, (and everything in `𝒜` has size `r`) then the initial segment we compare to
+is just all the subsets of `{0, ..., k - 1}` of size `r`. The `i`-th iterated shadow of this is all
+the subsets of `{0, ..., k - 1}` of size `r - i`, so the `i`-th iterated shadow of `𝒜` has at least
+`k.choose (r - i)` elements. -/
+theorem lovasz_form (hir : i ≤ r) (hrk : r ≤ k) (hkn : k ≤ n)
```
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/KruskalKatona.lean:301]: I'm not entirely convinced this is a great name either - this is a special case of the Lovasz form, and that's not even a particularly helpful theorem name. Certainly the docstring should mention that this is Lovasz' formulation and emphasise that over the Kruskal-Katona name, however.

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean [hunk 3]
```diff
@@ -290,5 +292,112 @@ theorem iterated_kk (h₁ : (𝒜 : Set (Finset (Fin n))).Sized r) (h₂ : 𝒞.
...
+    (h₁ : (𝒜 : Set (Finset (Fin n))).Sized r) (h₂ : k.choose r ≤ 𝒜.card) :
+    k.choose (r - i) ≤ (∂^[i] 𝒜).card := by
+  set range'k : Finset (Fin n) :=
+    attachFin (range k) fun m ↦ by rw [mem_range]; apply forall_lt_iff_le.2 hkn
+  set 𝒞 : Finset (Finset (Fin n)) := powersetCard r range'k
+  have Ccard : 𝒞.card = k.choose r := by rw [card_powersetCard, card_attachFin, card_range]
+  have : (𝒞 : Set (Finset (Fin n))).Sized r := Set.sized_powersetCard _ _
+  suffices this : (∂^[i] 𝒞).card = k.choose  (r - i) by
```
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/KruskalKatona.lean:309]:
```suggestion
  suffices this : (∂^[i] 𝒞).card = k.choose (r - i) by
```

File: Mathlib/Combinatorics/SetFamily/KruskalKatona.lean (hunk 2 above)
Comment (YaelDillies) [reply Mathlib/Combinatorics/SetFamily/KruskalKatona.lean]: Let me know how you like the new name and docstring

**Comment by b-mehta** [comment]
I wrote the original version of this many years ago so I won't hit merge, but it looks good to me.

maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**